File: questModule/src/main/java/ru/nekit/android/qls/quest/generator/TextQuestGenerator.java
package ru.nekit.android.qls.quest.generator;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;

import java.util.List;

import ru.nekit.android.qls.R;
import ru.nekit.android.qls.quest.IQuest;
import ru.nekit.android.qls.quest.QuestContext;
import ru.nekit.android.qls.quest.QuestionType;
import ru.nekit.android.qls.quest.types.TextQuest;
import ru.nekit.android.qls.utils.MathUtils;

public class TextQuestGenerator implements IQuestGenerator {

    private static final int DEFAULT_TEXT_CAMOUFLAGE_LENGTH = 4;
    private static final int DEFAULT_TEXT_CAMOUFLAGE_WORD_LENGTH = 5;
    private static final String DEFAULT_TEXT_CAMOUFLAGE_WORD = "слово";

    private QuestContext mQuestContext;
    private Type mType;
    private TextQuest mQuest;
    private int mCamouflageLength, mWordLength;

    public TextQuestGenerator(@NonNull QuestContext context, @Nullable QuestionType questionType) {
        mQuestContext = context;
        mQuest = new TextQuest(questionType);
    }

    private String makeNoisePattern(@NonNull Context context,
                                    @StringRes int supportCharsResourceIds[],
                                    int charCountBefore, int charCountAfter) {
        StringBuilder noisePattern = new StringBuilder();
        StringBuilder supportCharsBuilder = new StringBuilder();
        for (int resId : supportCharsResourceIds) {
            supportCharsBuilder.append(context.getString(resId));
        }
        String supportChars = supportCharsBuilder.toString();
        final int supportCharsLength = supportChars.length() - 1;
        int i = 0;
        char _char;
        for (; i < charCountBefore; i++) {
            _char = supportChars.charAt(MathUtils.randUnsignedInt(supportCharsLength));
            noisePattern.append(_char);
        }
        i = 0;
        noisePattern.append("%s");
        for (; i < charCountAfter; i++) {
            _char = supportChars.charAt(MathUtils.randUnsignedInt(supportCharsLength));
            noisePattern.append(_char);
        }
        return noisePattern.toString();
    }

    @Override
    public IQuest generate() {
        if (mType == Type.TEXT_CAMOUFLAGE) {
            List<String> wordList = mQuestContext.getQuestResourceLibrary().getWordList(mWordLength);
            String answer = DEFAULT_TEXT_CAMOUFLAGE_WORD;
            if (wordList.size() > 0) {
                answer = wordList.get(MathUtils.randUnsignedInt(wordList.size() - 1));
            }
            String[] questionStringArray = new String[answer.length()];
            for (int i = 0; i < answer.length(); i++) {
                char answerChar = answer.charAt(i);
                questionStringArray[i] = String.format(makeNoisePattern(mQuestContext,
                        new int[]{
                                R.string.text_quest_support_chars_ru,
                                R.string.text_quest_support_chars_number},
                        mCamouflageLength, 0), answerChar);
            }
            mQuest.questionStringArray = questionStringArray;
            mQuest.setAnswer(answer);
            return mQuest;
        }
        return null;
    }

    public void makeTextCamouflage(int wordLength, int camouflageLength) {
        mWordLength = wordLength;
        if (mWordLength == 0) {
            mWordLength = DEFAULT_TEXT_CAMOUFLAGE_WORD_LENGTH;
        }
        mCamouflageLength = camouflageLength;
        if (mCamouflageLength == 0) {
            mCamouflageLength = DEFAULT_TEXT_CAMOUFLAGE_LENGTH;
        }
        mType = Type.TEXT_CAMOUFLAGE;
    }

    private enum Type {
        TEXT_CAMOUFLAGE
    }
}

File: questPresentation/src/main/java/ru/nekit/android/qls/quest/generator/TextQuestGenerator.kt
package ru.nekit.android.qls.quest.generator

import ru.nekit.android.qls.quest.QuestContext
import ru.nekit.android.qls.shared.model.QuestionType

class TextQuestGenerator(private val mQuestContext: QuestContext, questionType: QuestionType?) /*: IQuestGenerator<TextQuest> */ {

    /*private var mType: Type? = null
    private val mQuest: TextQuest
    private var mCamouflageLength: Int = 0
    private var mWordLength: Int = 0

    init {
        mQuest = TextQuest()
        mQuest.questionType = questionType
    }

    private fun makeNoisePattern(context: android.content.Context,
                                 @StringRes supportCharsResourceIds: IntArray,
                                 charCountBefore: Int, charCountAfter: Int): String {
        val noisePattern = StringBuilder()
        val supportCharsBuilder = StringBuilder()
        for (resId in supportCharsResourceIds) {
            supportCharsBuilder.append(context.getString(resId))
        }
        val supportChars = supportCharsBuilder.to()
        val supportCharsLength = supportChars.length - 1
        var i = 0
        var _char: Char
        while (i < charCountBefore) {
            _char = supportChars[MathUtils.randUnsignedInt(supportCharsLength)]
            noisePattern.append(_char)
            i++
        }
        i = 0
        noisePattern.append("%s")
        while (i < charCountAfter) {
            _char = supportChars[MathUtils.randUnsignedInt(supportCharsLength)]
            noisePattern.append(_char)
            i++
        }
        return noisePattern.to()
    }

    override fun generate(): TextQuest {
        if (mType == Type.TEXT_CAMOUFLAGE) {
            val wordList = mQuestContext.questResourceRepository.getWordList(mWordLength)
            var answer = DEFAULT_TEXT_CAMOUFLAGE_WORD
            if (wordList.isNotEmpty()) {
                answer = wordList[MathUtils.randUnsignedInt(wordList.size - 1)]
            }
            val questionStringArray = arrayOfNulls<String>(answer.length)
            for (i in 0 until answer.length) {
                val answerChar = answer[i]
                questionStringArray[i] = String.format(makeNoisePattern(mQuestContext,
                        intArrayOf(R.string.text_quest_support_chars_ru, R.string.text_quest_support_chars_number),
                        mCamouflageLength, 0), answerChar)
            }
            mQuest.questionStringArray = questionStringArray
            mQuest.setAnswer(answer)
            return mQuest
        }
        return null
    }

    fun makeTextCamouflage(wordLength: Int, camouflageLength: Int) {
        mWordLength = wordLength
        if (mWordLength == 0) {
            mWordLength = DEFAULT_TEXT_CAMOUFLAGE_WORD_LENGTH
        }
        mCamouflageLength = camouflageLength
        if (mCamouflageLength == 0) {
            mCamouflageLength = DEFAULT_TEXT_CAMOUFLAGE_LENGTH
        }
        mType = Type.TEXT_CAMOUFLAGE
    }

    private enum class Type {
        TEXT_CAMOUFLAGE
    }

    companion object {

        private val DEFAULT_TEXT_CAMOUFLAGE_LENGTH = 4
        private val DEFAULT_TEXT_CAMOUFLAGE_WORD_LENGTH = 5
        private val DEFAULT_TEXT_CAMOUFLAGE_WORD = "слово"
    }*/
}
